File: assignments/clojure/allergies/allergies_test.clj
(ns allergies.test (:require [clojure.test :refer :all]))
(load-file "allergies.clj")

(deftest no-allergies-at-all
  (is (= [] (allergies/list 0))))

(deftest allergic-to-just-eggs
  (is (= ["eggs"] (allergies/list 1))))

(deftest allergic-to-just-peanuts
  (is (= ["peanuts"] (allergies/list 2))))

(deftest allergic-to-just-strawberries
  (is (= ["strawberries"] (allergies/list 8))))

(deftest allergic-to-eggs-and-peanuts
  (is (= ["eggs", "peanuts"] (allergies/list 3))))

(deftest allergic-to-more-than-eggs-but-not-peanuts
  (is (= ["eggs", "shellfish"] (allergies/list 5))))

(deftest allergic-to-lots-of-stuff
  (is (= ["strawberries", "tomatoes", "chocolate", "pollen", "cats"] (allergies/list 248))))

(deftest allergic-to-everything
  (is (= ["eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"] (allergies/list 255))))

(deftest no-allergies-means-not-allergic
  (is (not (allergies/allergic_to? 0 "peanuts")))
  (is (not (allergies/allergic_to? 0 "cats")))
  (is (not (allergies/allergic_to? 0 "strawberries"))))

(deftest is-allergic-to-eggs
  (is (allergies/allergic_to? 1 "eggs")))

(deftest allergic-to-eggs-in-addition-to-other-stuff
  (is (allergies/allergic_to? 5 "eggs")))

(deftest ignore-non-allergen-score-parts
  (is (= ["eggs", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"] (allergies/list 509))))

(run-tests)

File: assignments/haskell/allergies/allergies_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Allergies (Allergen(..), isAllergicTo, allergies)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestList allergiesTests ]

allergiesTests :: [Test]
allergiesTests =
  [ testCase "no allergies at all" $
    [] @=? allergies 0
  , testCase "allergic to just eggs" $
    [Eggs] @=? allergies 1
  , testCase "allergic to just peanuts" $
    [Peanuts] @=? allergies 2
  , testCase "allergic to just strawberries" $
    [Strawberries] @=? allergies 8
  , testCase "allergic to eggs and peanuts" $
    [Eggs, Peanuts] @=? allergies 3
  , testCase "allergic to more than eggs but not peanuts" $
    [Eggs, Shellfish] @=? allergies 5
  , testCase "allergic to lots of stuff" $
    [Strawberries, Tomatoes, Chocolate, Pollen, Cats] @=? allergies 248
  , testCase "allergic to everything" $
    [ Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate
    , Pollen, Cats ] @=? allergies 255
  , testCase "no allergies means not allergic" $ do
    False @=? isAllergicTo Peanuts 0
    False @=? isAllergicTo Cats 0
    False @=? isAllergicTo Strawberries 0
  , testCase "is allergic to eggs" $
    True @=? isAllergicTo Eggs 1
  , testCase "allergic to eggs in addition to other stuff" $
    True @=? isAllergicTo Eggs 5
  , testCase "ignore non allergen score parts" $
    [ Eggs, Shellfish, Strawberries, Tomatoes, Chocolate
    , Pollen, Cats ] @=? allergies 509
  ]
